sky/packages/sky/lib/gestures/fling.dart

```diff
@@ -0,0 +1,63 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:sky' as sky;
+
+import 'package:sky/gestures/arena.dart';
+import 'package:sky/gestures/recognizer.dart';
+import 'package:sky/gestures/constants.dart';
+
+// Fling velocities are logical pixels per second.
+typedef void GestureFlingCallback(sky.Offset velocity);
+
+int _eventTime(sky.PointerEvent event) => (event.timeStamp * 1000.0).toInt(); // microseconds
+
+bool _isFlingGesture(sky.GestureVelocity velocity) {
+  double vSquared = velocity.x * velocity.x + velocity.y * velocity.y;
+  return velocity.isValid &&
+    vSquared > kMinFlingVelocity * kMinFlingVelocity &&
+    vSquared < kMaxFlingVelocity * kMaxFlingVelocity;
+}
+
+class FlingGestureRecognizer extends GestureRecognizer {
+  FlingGestureRecognizer({ PointerRouter router, this.onFling })
+    : super(router: router);
+
+  GestureFlingCallback onFling;
+  sky.VelocityTracker velocityTracker = new sky.VelocityTracker();
+  int primaryPointer;
+  int lastEventTime;
+
+  void addPointer(sky.PointerEvent event) {
+    startTrackingPointer(event.pointer);
+    if (primaryPointer == null)
+      primaryPointer = event.pointer;
+  }
+
+  void handleEvent(sky.PointerEvent event) {
+    if (event.pointer == primaryPointer) {
+      if (event.type == 'pointermove') {
+        velocityTracker.addPosition(_eventTime(event), primaryPointer, event.x, event.y);
+      } else if (event.type == 'pointerup') {
+        sky.GestureVelocity velocity = velocityTracker.getVelocity(primaryPointer);
+        if (_isFlingGesture(velocity)) {
+          resolve(GestureDisposition.accepted);
+          this?.onFling(new sky.Offset(velocity.x, velocity.y));
```

Comment:
`this` can't ever be null.  We can skip the ?

I don't think there's a clever way to check onFling for null, but I think it would be fair to assert that onFling is non-null here and in the constructor.